File: www/v2/adapter/SyncActionAdapter.php
<?php

class SyncActionAdapter extends DbAdapter {
	
	private $adapterCache;
	
	public function __construct() {
		parent::__construct();
		
		$this->adapterCache = array();
		$this->adapterCache["Words"] = new WordAdapter();
		$this->adapterCache["Memos"] = new MemoAdapter();
		$this->adapterCache["MemoBases"] = new MemoBaseAdapter();
	}
	
	public function syncServer($syncClientActions) {
		$dbTran = parent::transConnect();
		$dbTran->beginTransaction();
		
		try {
            
            // If new item is created and updated afterwards, only update is sent
            // however, multiple updates may bey sent, but only one insert should be performed.
            $updateInsertContext = array();
			
			foreach($syncClientActions as $syncAction) {
				$adapter = $this->getAdapter($syncAction);
				
				switch ($syncAction->Action) {
					case SyncAction::ACTION_DELETE:
                        // One action per key
						$adapter->deleteEntity($dbTran, $syncAction->PrimaryKey, $syncAction);
						break;
					case SyncAction::ACTION_INSERT:
                        // One action per key
						$object = $syncAction->getClientObject($adapter);
						$adapter->insertEntity($dbTran, $object, $syncAction);
						break;
					case SyncACtion::ACTION_UPDATE:
                        // Multiple actions per key
                        $exists = in_array($syncAction->Table . $syncAction->PrimaryKey, $updateInsertContext) 
                            || $adapter->getEntity($syncAction->PrimaryKey) != null;
						$object = $syncAction->getClientObject($adapter);
                        
                        if(!$exists) { 
                            $adapter->insertEntity($dbTran, $object, $syncAction);
                            $updateInsertContext[] = $syncAction->Table . $syncAction->PrimaryKey;
                        } else {
						  $adapter->updateEntity($dbTran, $object, $syncAction);
                        }
                    
						break;
				}
			}
			
		} catch(Exception $ex) {
			$dbTran->rollBack();
			throw $ex;
		}
		
		$dbTran->commit();
	}
	
	// Returns SyncPackage
	public function syncClient($syncClientId, $syncServerActions) {

		$syncPackage = new SyncPackage();
		
		$syncObjects = array();
		$index = 0;
		
		$syncActionsToDiscard = array();
		foreach($syncServerActions as $syncAction) {
			$adapter = $this->getAdapter($syncAction);
			
			if($syncAction->Action != SyncAction::ACTION_DELETE) {
				$object = $adapter->getEntity($syncAction->PrimaryKey);
				
				// If object is null, this means that during sync
				// Some objects has been removed - e.g. Children of deleted Parent
				// Resolver does not know about the foreign key relationship
				// so we need to enforce it here
				
				if($object == null) {
					$syncActionsToDiscard[] = $syncAction;
					continue;
				}
				
				$key = array_search($object, $syncObjects);
				if($key == false) {
					$key = $index++;
				}
				$syncObjects[$key] = $object;
				$syncAction->SyncObjectId = $key;
			}
		}
		
		foreach($syncActionsToDiscard as $actionKey=>$action) {
			unset($syncServerActions[$actionKey]);
		}
        
		$syncPackage->SyncClientId = $syncClientId;
		$syncPackage->SyncActions = $syncServerActions;
		$syncPackage->ServerObjects = $syncObjects;
		
		return $syncPackage;
	}
	
	public static function insertAction($dbTran, $syncAction) {
		if($syncAction->SyncClientId == null) {
			// Not tracked one
			return;
		}
		
		$dbTran->beginTransaction();
		
		$existingRecords = self::getSimilar($dbTran, $syncAction->SyncClientId, $syncAction->Table, $syncAction->PrimaryKey, $syncAction->UpdateColumn);

		// Delete all existing actions add new insert action
		foreach($existingRecords as $existingRecord) {
			self::deleteDb($dbTran, $existingRecord->SyncActionId);
		}
		
		self::insertDb($dbTran, $syncAction);
		
		$dbTran->commit();
	}
	
	public static function deleteAction($dbTran, $syncAction) {
		if($syncAction->SyncClientId == null) {
			// Not tracked one
			return;
		}
		
		$dbTran->beginTransaction();
		
		$existingRecords = self::getSimilar($dbTran, $syncAction->SyncClientId, $syncAction->Table, $syncAction->PrimaryKey, $syncAction->UpdateColumn);
		
		// Delete all existing actions add new delete action
		foreach($existingRecords as $existingRecord) {
			self::deleteDb($dbTran, $existingRecord->SyncActionId);
		}

		self::insertDb($dbTran, $syncAction);

		$dbTran->commit();
	}
	
	public static function updateAction($dbTran, $syncAction) {
		if($syncAction->SyncClientId == null) {
			// Not tracked one
			return;
		}
		
		$dbTran->beginTransaction();
		
		$existingRecords = self::getSimilar($dbTran, $syncAction->SyncClientId, $syncAction->Table, $syncAction->PrimaryKey, $syncAction->UpdateColumn);
		
		// Delete all existing actions add new update action
		foreach($existingRecords as $existingRecord) {
			self::deleteDb($dbTran, $existingRecord->SyncActionId);
		}

		self::insertDb($dbTran, $syncAction);

		$dbTran->commit();
	}
	
	public static function removeActions($dbTran, $table, $primaryKey) {
		$dbTran->beginTransaction();
		
		$query = "DELETE FROM memoling_SyncActions WHERE ". 
				 "`Table` = :Table AND ".
				 "PrimaryKey = :PrimaryKey";
		
		$stm = $dbTran->prepare($query);
		$stm->bindParam("Table", $table);
		$stm->bindParam("PrimaryKey", $primaryKey);
		
		if(!$stm->execute()) {
			$dbTran->rollBack();
			throw new SqlException("Failed to remove sync actions from DB", $dbTran);
		}
		
		$dbTran->commit();
	}
	
	public function get($syncClientId, $timestmap) {
		return self::getDb(parent::connect(), $syncClientId, $timestmap);
	}
	
	public static function getDb($db, $syncClientId, $timestamp) {
	
		$stm = $db->prepare("SELECT * FROM memoling_SyncActions WHERE SyncClientId = :SyncClientId AND ServerTimestamp >= :ServerTimestamp ORDER BY ServerTimestamp");
		$stm->bindParam("SyncClientId", $syncClientId);
		$stm->bindParam("ServerTimestamp", $timestamp);
		$stm->execute();
	
		$array = array();
		while($row = $stm->fetch()) {
			$array[] = self::bindSyncAction($row);
		}
	
		return $array;
	}
	
	private function getAdapter($syncAction) {
		return $this->adapterCache[$syncAction->Table];
	}
	
	private static function getSimilar($dbTran, $syncClientId, $table, $primaryKey, $updateColumn) {
	
		$query = "SELECT * FROM memoling_SyncActions WHERE SyncClientId = :SyncClientId "
				."AND `Table` = :Table AND PrimaryKey = :PrimaryKey ";
	
		if($updateColumn != null) {
			$query .= "AND (UpdateColumn = :UpdateColumn OR UpdateColumn IS NULL) ";
		}
	
		$query .= "ORDER BY ServerTimestamp DESC";
	
		$stm = $dbTran->prepare($query);
		$stm->bindParam("SyncClientId", $syncClientId);
		$stm->bindParam("Table", $table);
		$stm->bindParam("PrimaryKey", $primaryKey);
		
		if($updateColumn != null) {
			$stm->bindParam("UpdateColumn", $updateColumn);
		}
		
		$stm->execute();
	
		$array = array();
		while($row = $stm->fetch()) {
			$array[] = self::bindSyncAction($row);
		}
	
		return $array;
	}
	
	private function insert($action) {
		return self::insertDb(parent::connect(), $action);
	}
	
	private static function insertDb($dbTran, $action) {
		$query = "INSERT INTO
				memoling_SyncActions(SyncActionId,SyncClientId,`Table`,PrimaryKey,Action,ServerTimestamp,UpdateColumn)
				VALUES(:SyncActionId,:SyncClientId,:Table,:PrimaryKey,:Action,:ServerTimestamp,:UpdateColumn)";
	
		$stm = $dbTran->prepare($query);
		$stm->bindParam("SyncActionId", $action->SyncActionId);
		$stm->bindParam("SyncClientId", $action->SyncClientId);
		$stm->bindParam("Table", $action->Table);
		$stm->bindParam("PrimaryKey", $action->PrimaryKey);
		$stm->bindParam("Action", $action->Action, PDO::PARAM_INT);
		$stm->bindParam("ServerTimestamp", $action->ServerTimestamp);
		$stm->bindParam("UpdateColumn", $action->UpdateColumn);
	
		if(!$stm->execute()) {
			throw new SqlException("Failed to create new SyncAction", $dbTran);
		}
	}
	
	private static function deleteDb($dbTran, $syncActionId) {
		
		$query = "DELETE FROM
				memoling_SyncActions
				WHERE
				SyncActionId = :SyncActionId
				";
		
		$stm = $dbTran->prepare($query);
		$stm->bindParam(":SyncActionId", $syncActionId);
		if(!$stm->execute()) {
			throw new SqlException("Failed to delete DeletWord", $dbTran);
		}
	}
	
	private static function bindSyncAction($row) {
		$obj = new SyncAction();
	
		$obj->SyncActionId = $row["SyncActionId"];
		$obj->SyncClientId = $row["SyncClientId"];
		$obj->Table = $row["Table"];
		$obj->Action = intval($row["Action"]);
		$obj->PrimaryKey = $row["PrimaryKey"];
		$obj->ServerTimestamp = $row["ServerTimestamp"];
		$obj->UpdateColumn  = $row["UpdateColumn"];
	
		return $obj;
	}
}

?>

File: android/Memoling/src/app/memoling/android/adapter/SyncActionAdapter.java
package app.memoling.android.adapter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.json.JSONException;
import org.json.JSONObject;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import app.memoling.android.db.DatabaseHelper;
import app.memoling.android.db.SqliteAdapter;
import app.memoling.android.entity.SyncAction;
import app.memoling.android.helper.AppLog;
import app.memoling.android.sync.cloud.ISyncAdapter;
import app.memoling.android.sync.cloud.ISyncEntity;
import app.memoling.android.sync.cloud.SyncPackage;


public class SyncActionAdapter extends SqliteAdapter {

	public final static String TableName = "SyncActions";
	
	private Map<String,ISyncAdapter> m_adapterCache;
	
	public SyncActionAdapter(Context context) {
		super(context);
		
		m_adapterCache = new HashMap<String,ISyncAdapter>();
		
		m_adapterCache.put("Words", new WordAdapter(context));
		m_adapterCache.put("Memos", new MemoAdapter(context));
		m_adapterCache.put("MemoBases", new MemoBaseAdapter(context));
	}
	
	public boolean syncClient(List<SyncAction> syncServerActions) {

		SQLiteDatabase db = getDatabase();
		db.beginTransaction();
		
		try {
		
			// Remove old actions
			SyncActionAdapter.removeActions(db);
			
			// If new item is created and updated afterwards, only update is sent
			// however, multiple updates may be sent, but only one insert is allowed.
			List<String> updateInsertContext = new ArrayList<String>();
			
			// Sync all actions
			for(SyncAction syncAction : syncServerActions) {
				ISyncAdapter adapter = getAdapter(syncAction);
				
				switch(syncAction.getAction()) {
					case SyncAction.ACTION_DELETE:
						// One action per key
						adapter.deleteEntity(db, syncAction.getPrimaryKey(), syncAction);
						break;
					case SyncAction.ACTION_INSERT:
						// One action per key
						ISyncEntity insertObject = syncAction.getServerObject(adapter);
						adapter.insertEntity(db, insertObject, syncAction);
						break;
					case SyncAction.ACTION_UPDATE:
						// Multiple actions per key
						boolean exists = updateInsertContext.contains(syncAction.getTable() + syncAction.getPrimaryKey())
							|| adapter.getEntity(syncAction.getPrimaryKey()) != null;
						ISyncEntity updateObject = syncAction.getServerObject(adapter);
						
						if(!exists) {
							adapter.insertEntity(db, updateObject, syncAction);
							updateInsertContext.add(syncAction.getTable() + syncAction.getPrimaryKey());
						} else {
							adapter.updateEntity(db, updateObject, syncAction);	
						}
						
						break;
				}
				
			}

			// Remove new actions (from server)
			SyncActionAdapter.removeActions(db);
	
			// Update last sync timestamp
			SyncClientAdapter.updateLastSyncServerTimestamp(db, SyncClientAdapter.getCurrentSyncClientId(db));
			
			db.setTransactionSuccessful();
			
			return true;
		} catch(Exception ex) {
			AppLog.e("SyncActionAdapter", "syncClient - Sync Error", ex);
			return false;
		} finally {
			db.endTransaction();
		}

	}
	
	public SyncPackage syncServer(List<SyncAction> syncActions) {
		SyncPackage syncPackage = new SyncPackage();
		
		syncPackage.setSyncActions(syncActions);
		HashMap<String, JSONObject> clientObjects = new HashMap<String, JSONObject>();
		int index = 0;
		
		// Get all actions
		for(SyncAction syncAction : syncActions) {
			// Add objects if not deleted 
			if(syncAction.getAction() != SyncAction.ACTION_DELETE) {
				ISyncAdapter syncAdapter = getAdapter(syncAction);
				ISyncEntity clientObject = syncAdapter.getEntity(syncAction.getPrimaryKey());
				syncAction.setSyncObjectId(Integer.toString(index));
				try {
					clientObjects.put(Integer.toString(index), clientObject.encodeEntity());
				} catch(JSONException ex) {
					AppLog.e("SyncActionAdapter", "SyncServer: Parse Exception", ex);
				}
			}
			index++;
		}
		
		syncPackage.setClientObjects(clientObjects);
		
		return syncPackage;
	}
	
	public void buildInitialSync() throws SQLiteException {
		
		// Add every record as insert in order MemoBase,Word,Memo
		// Order is determines by foreign key constraints

		try {
			SQLiteDatabase db = getDatabase();
			String syncClientId = SyncClientAdapter.getCurrentSyncClientId(db);
			
			// Clear all actions in the database 
			// Old ones might be due to unsuccessful initial syncs
			SyncActionAdapter.removeActions(db);

			m_adapterCache.get("MemoBases").buildInitialSync(db, syncClientId);
			m_adapterCache.get("Words").buildInitialSync(db, syncClientId);
			m_adapterCache.get("Memos").buildInitialSync(db, syncClientId);
		} finally {
			closeDatabase();
		}
	}
	
	private ISyncAdapter getAdapter(SyncAction syncAction) {
		
		String table = syncAction.getTable();
		
		if(table.equals("Words")) {
			return m_adapterCache.get("Words");
		} else if(table.equals("Memos")) {
			return m_adapterCache.get("Memos");
		} else if(table.equals("MemoBases")) {
			return m_adapterCache.get("MemoBases");
		}
		
		return null;
	}
	
	public static void insertAction(SQLiteDatabase db, SyncAction syncAction) {
		if(syncAction.getSyncClientId() == null) {
			// Not tracked record
			return;
		}
		
		db.beginTransaction();
		try {
			List<SyncAction> existingRecords = SyncActionAdapter.getSimilar(db, syncAction.getSyncClientId(), syncAction.getTable(), syncAction.getPrimaryKey(), syncAction.getUpdateColumn());
			
			// Delete all existing records and add insert action
			for(SyncAction existingRecord : existingRecords) {
				SyncActionAdapter.delete(db, existingRecord.getSyncActionId());
			}
			
			ContentValues values = createValues(syncAction);
			db.insertOrThrow(TableName, null, values);
			
			db.setTransactionSuccessful();
		} finally {
			db.endTransaction();
		}	
	}
	
	public static void deleteAction(SQLiteDatabase db, SyncAction syncAction) {
		if(syncAction.getSyncClientId() == null) {
			// Not tracked record
			return;
		}
	
		db.beginTransaction();
		try {
			List<SyncAction> existingRecords = SyncActionAdapter.getSimilar(db, syncAction.getSyncClientId(), syncAction.getTable(), syncAction.getPrimaryKey(), syncAction.getUpdateColumn());
			
			// Delete all existing records and add delete action
			for(SyncAction existingRecord : existingRecords) {
				SyncActionAdapter.delete(db, existingRecord.getSyncActionId());
			}
			
			ContentValues values = createValues(syncAction);
			db.insertOrThrow(TableName, null, values);
			
			db.setTransactionSuccessful();
		} finally {
			db.endTransaction();
		}
	}
	
	public static void updateAction(SQLiteDatabase db, SyncAction syncAction) {
		if(syncAction.getSyncClientId() == null) {
			// Not tracked record
			return;
		}
	
		db.beginTransaction();
		try {
			List<SyncAction> existingRecords = SyncActionAdapter.getSimilar(db, syncAction.getSyncClientId(), syncAction.getTable(), syncAction.getPrimaryKey(), syncAction.getUpdateColumn());
			
			// Delete all existing records and add update action
			for(SyncAction existingRecord : existingRecords) {
				SyncActionAdapter.delete(db, existingRecord.getSyncActionId());
			}
			
			ContentValues values = createValues(syncAction);
			db.insertOrThrow(TableName, null, values);
			
			db.setTransactionSuccessful();
		} finally {
			db.endTransaction();
		}
	}
	
	public static void removeAction(SQLiteDatabase db, String table, String primaryKey) {		
		db.beginTransaction();
		try {
			db.delete(TableName, "\"Table\" = ? AND PrimaryKey = ?", new String[] { table, primaryKey });
			db.setTransactionSuccessful();
		} finally {
			db.endTransaction();
		}
	}
	
	public List<SyncAction> get(String syncClientId, long serverTimestamp) {
		return SyncActionAdapter.get(getDatabase(), syncClientId, serverTimestamp);
	}
	
	public static List<SyncAction> get(SQLiteDatabase db, String syncClientId, long serverTimestamp) {
		Cursor cursor = null;
		ArrayList<SyncAction> syncActions = new ArrayList<SyncAction>();
		
		String sql = "SELECT * FROM SyncActions WHERE SyncClientId = ? AND ServerTimestamp >= ? ORDER BY ServerTimestamp";
		
		try {
			cursor = db.rawQuery(sql, new String[] { syncClientId, Long.toString(serverTimestamp) });
			
			while(cursor.moveToNext()) {
				SyncAction syncAction = SyncActionAdapter.bindSyncAction(cursor);
				syncActions.add(syncAction);
			}
		}
		finally {
			if(cursor != null) {
				cursor.close();
			}
		}
		
		return syncActions;
	}
	
	public List<SyncAction> getAll() {
		try {
			return SyncActionAdapter.getAll(getDatabase());
		} finally {
			closeDatabase();
		}
	}
	
	public static List<SyncAction> getAll(SQLiteDatabase db) {
		ArrayList<SyncAction> syncActions = new ArrayList<SyncAction>();
		Cursor cursor = null;
		
		// Get all actions in right order
		try {
			String sql = "SELECT * FROM SyncActions ORDER BY ServerTimestamp DESC";
			cursor = db.rawQuery(sql, null);
			while(cursor.moveToNext()) {
				SyncAction syncAction = bindSyncAction(cursor);
				syncActions.add(syncAction);
			}
			
			return syncActions;
		} finally {
			if(cursor != null) {
				cursor.close();
			}
		}
	}
	
	public void insert(SyncAction syncAction) {
		SyncActionAdapter.insert(getDatabase(), syncAction);
	}
	
	public static void insert(SQLiteDatabase db, SyncAction syncAction) {
		db.beginTransaction();
		
		try {
			ContentValues values = SyncActionAdapter.createValues(syncAction);
			db.insertOrThrow(TableName, null, values);
			db.setTransactionSuccessful();
		} finally {
			db.endTransaction();
		}
	}
	
	public void delete(String syncActionId) {
		try {
			SyncActionAdapter.delete(getDatabase(), syncActionId);
		} finally {
			closeDatabase();
		}
		
	}

	private static void delete(SQLiteDatabase db, String syncActionId) {
		db.delete(TableName, "SyncActionId = ?", new String[] { syncActionId });
	}
	
	private static void removeActions(SQLiteDatabase db) {
		db.delete(TableName, null, null);
	}
	
	private static List<SyncAction> getSimilar(SQLiteDatabase db, String syncClientId, String table, String primaryKey, String updateColumn) {
		
		String sql = "SELECT * FROM SyncActions WHERE SyncClientId = ? AND \"Table\" = ? AND PrimaryKey = ?";
		int argsCount = 3;
		
		if(updateColumn != null) {
			sql += " AND (UpdateColumn = ? OR UpdateColumn IS NULL)";
			argsCount++;
		}

		String[] args = new String[argsCount];
		args[0] = syncClientId;
		args[1] = table;
		args[2] = primaryKey;
		if(argsCount > 3) {
			args[3] = updateColumn;
		}
		
		sql += " ORDER BY ServerTimestamp DESC";
		
		Cursor cursor = null;
		try {
			
			cursor = db.rawQuery(sql, args);
			ArrayList<SyncAction> syncActions = new ArrayList<SyncAction>();
			while(cursor.moveToNext()) {
				SyncAction syncAction = SyncActionAdapter.bindSyncAction(cursor);
				syncActions.add(syncAction);
			}
			
			return syncActions;
		} finally {
			if(cursor != null) {
				cursor.close();
			}
		}
	}
	
	private static SyncAction bindSyncAction(Cursor cursor) {
		SyncAction syncAction = new SyncAction();
		
		syncAction.setAction(DatabaseHelper.getInt(cursor, "Action"));
		syncAction.setPrimaryKey(DatabaseHelper.getString(cursor, "PrimaryKey"));
		syncAction.setServerTimestamp(DatabaseHelper.getLong(cursor, "ServerTimestamp"));
		syncAction.setSyncClientId(DatabaseHelper.getString(cursor, "SyncClientId"));
		syncAction.setTable(DatabaseHelper.getString(cursor, "Table"));
		syncAction.setUpdateColumn(DatabaseHelper.getString(cursor, "UpdateColumn"));
		syncAction.setSyncActionId(DatabaseHelper.getString(cursor, "SyncActionId"));
		
		return syncAction;
	}
	
	private static ContentValues createValues(SyncAction syncAction) {
		ContentValues values = new ContentValues();
		values.put("Action", syncAction.getAction());
		values.put("PrimaryKey", syncAction.getPrimaryKey());
		values.put("ServerTimestamp", syncAction.getServerTimestamp());
		values.put("SyncClientId", syncAction.getSyncClientId());
		values.put("\"Table\"", syncAction.getTable());
		values.put("UpdateColumn", syncAction.getUpdateColumn());
		values.put("SyncClientId", syncAction.getSyncClientId());
		values.put("SyncActionId", syncAction.getSyncActionId());
		return values;
	}
}
